viewer-prototype/src/browser/trace-viewer/components/utils/filtrer-tree/tree.tsx

```diff
@@ -0,0 +1,186 @@
+        const checkedNode = this.getNode(id);
+        if (checkedNode) {
+            if (checkedNode.children.length) {
+                const childrenIds = this.getAllChildrenIds(checkedNode, []);
+                if (this.isNodeChecked(id)) {
+                    checkedIds = checkedIds.concat(childrenIds)
+                } else {
+                    const childIdsToCheck = childrenIds.filter(id => {
+                        return !this.isNodeChecked(id);
+                    });
+                    checkedIds = checkedIds.concat(childIdsToCheck);
+                }
+            } else {
+                checkedIds.push(id);
+                if (this.areAllSiblingsChecked(checkedNode)) {
+                    checkedIds.push(checkedNode.parentId);
+                }
+            }
+            this.props.onChecked(checkedIds);
+        }
+    }
+
+    //returns 0 for unchecked, 1 for checked, 2 for half checked
+    getCheckedStatus = (id: number): number => {
+        const node = this.getNode(id);
+        if (node) {
+            if (node.children.length === 0) {
+                return this.isNodeChecked(id) ? 1 : 0;
+            } else {
+                if (this.isEveryChildChecked(node)) {
+                    return 1;
+                }
+                if (this.isSomeChildChecked(node)) {
+                    return 2;
+                }
+            }
+        }
+        return 0;
+    }
+
+    isEveryChildChecked = (node: TreeNode): boolean => {
+        return node.children.every((child: TreeNode) => {
+            return this.isNodeChecked(child.id);
+        })
+    }
+
+    isSomeChildChecked = (node: TreeNode): boolean => {
+        return node.children.some((child: TreeNode) => {
+            return this.isNodeChecked(child.id);
+        })
+    }
+
+    renderTreeNodes = (nodes: TreeNode[], parent: TreeNode = defaultTreeNode, level: number = 0): JSX.Element | null => {
+        const treeNodes = nodes.map((node: TreeNode) => {
+            const children = node.children.length > 0 ? this.renderTreeNodes(node.children, node, level+1) : null;
+            const checkedStatus = this.getCheckedStatus(node.id);
+
+            if (!parent.expanded) {
+                return null;
+            }
+            return (
+                <TreeNodeComponent 
+                    node={node}
+                    key={'node-'+node.id}
+                    level={level}
+                    checkedStatus={checkedStatus}
+                    isCheckable={this.props.showCheckboxes}
+                    onExpanded={this.handleExpand}
+                    onChecked={this.handleCheck}
+                >
+                    {children}
+                </TreeNodeComponent>
+            )
+        });
+        return (
+            <React.Fragment>
+                {treeNodes}
+            </React.Fragment>
+        )
+    }
+
+    render() {
+        if (!this.props.nodes || !this.props.checkedSeries) return null;
```

Comment:
You said checkedSeries should be optional above, this check is making it mandatory?